plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncUploadUtils.java

```diff
@@ -0,0 +1,464 @@
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+/**
+ * A helper class that automatically uses multipart upload based on the size of the source object
+ */
+public final class AsyncUploadUtils {
+    private static final Logger log = LogManager.getLogger(AsyncUploadUtils.class);
+    private final ExecutorService executorService;
+    private final ExecutorService priorityExecutorService;
+    private final long minimumPartSize;
+
+    /**
+     * The max number of parts on S3 side is 10,000
+     */
+    private static final long MAX_UPLOAD_PARTS = 10_000;
+
+    /**
+     * Construct a new object of AsyncUploadUtils
+     *
+     * @param minimumPartSize         The minimum part size for parallel multipart uploads
+     * @param executorService         The stream reader {@link ExecutorService} for normal priority uploads
+     * @param priorityExecutorService the stream read {@link ExecutorService} for high priority uploads
+     */
+    public AsyncUploadUtils(long minimumPartSize, ExecutorService executorService, ExecutorService priorityExecutorService) {
+        this.executorService = executorService;
+        this.priorityExecutorService = priorityExecutorService;
+        this.minimumPartSize = minimumPartSize;
+    }
+
+    /**
+     * Upload an object to S3 using the async client
+     *
+     * @param s3AsyncClient The {@link S3AsyncClient} to use for uploads
+     * @param uploadRequest The {@link UploadRequest} object encapsulating all relevant details for upload
+     * @param streamContext The {@link StreamContext} to supply streams during upload
+     * @return A {@link CompletableFuture} to listen for upload completion
+     */
+    public CompletableFuture<Void> uploadObject(S3AsyncClient s3AsyncClient, UploadRequest uploadRequest, StreamContext streamContext) {
+
+        CompletableFuture<Void> returnFuture = new CompletableFuture<>();
+        try {
+            if (streamContext.getNumberOfParts() == 1) {
+                log.debug(() -> "Starting the upload as a single upload part request");
+                uploadInOneChunk(s3AsyncClient, uploadRequest, streamContext.provideStream(0), returnFuture);
+            } else {
+                log.debug(() -> "Starting the upload as multipart upload request");
+                uploadInParts(s3AsyncClient, uploadRequest, streamContext, returnFuture);
+            }
+        } catch (Throwable throwable) {
+            returnFuture.completeExceptionally(throwable);
+        }
+
+        return returnFuture;
+    }
```

Comment:
I quickly tried moving the CompletableFuture to a Listener. Let me know what you think
```
-    public CompletableFuture<Void> uploadObject(S3AsyncClient s3AsyncClient, UploadRequest uploadRequest, StreamContext streamContext) {
-
+    public void uploadObject(UploadRequest uploadRequest, StreamContext streamContext, ActionListener<Void> completionListener) {
         CompletableFuture<Void> returnFuture = new CompletableFuture<>();
         try {
             if (streamContext.getNumberOfParts() == 1) {
@@ -99,11 +100,10 @@ public final class AsyncUploadUtils {
                 log.debug(() -> "Starting the upload as multipart upload request");
                 uploadInParts(s3AsyncClient, uploadRequest, streamContext, returnFuture);
             }
-        } catch (Throwable throwable) {
-            returnFuture.completeExceptionally(throwable);
+            returnFuture.whenComplete(ActionListener.map(ActionListener.toBiConsumer(completionListener)));
+        } catch (Exception exception) {
+            completionListener.onFailure(exception);
         }
-
-        return returnFuture;
     }
 
     private void uploadInParts(
diff --git a/server/src/main/java/org/opensearch/action/ActionListener.java b/server/src/main/java/org/opensearch/action/ActionListener.java
index 645ed4deec0..ac07bc1d304 100644
--- a/server/src/main/java/org/opensearch/action/ActionListener.java
+++ b/server/src/main/java/org/opensearch/action/ActionListener.java
@@ -217,6 +217,24 @@ public interface ActionListener<Response> {
         };
     }
 
+    /**
+     ** Converts a Listener that supports Exception to Throwable
+     * @param listener
+     * @param <Response>
+     * @return
+     */
+    static <Response> BiConsumer<Response, Throwable> map(BiConsumer<Response, ? super Exception> listener) {
+        BiConsumer<Response, Throwable> castThrowable = (v, t) -> {
+            if (t == null) {
+                listener.accept(v, null);
+            } else {
+                assert !(t instanceof Error) : "Cannot be error";
+                listener.accept(v, (Exception) t);
+            }
+        };
+        return castThrowable;
+    }
+

```